gh-105375: Improve error handling in _Unpickler_SetInputStream()

erlend-aasland opened the pull request:
Prevent exceptions from possibly being overwritten in case of multiple
failures.


<!-- gh-issue-number: gh-105375 -->
* Issue: gh-105375
<!-- /gh-issue-number -->

erlend-aasland committed:
gh-105375: Improve error handling in _Unpickler_SetInputStream()

Prevent exceptions from possibly being overwritten in case of multiple
failures.

erlend-aasland committed:
Add NEWS

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -1678,25 +1678,40 @@ _Unpickler_SetInputStream(UnpicklerObject *self, PyObject *file)
+        PyErr_Clear();
+        no_readline_attr = 1;
+    }
+    if (no_read_attr || no_readline_attr) {
```
Why add `no_read_attr` and `no_readline_attr`? Why not simply check `self->readline` and `self->read`?

erlend-aasland replied:
True, the added variables are not needed; I should have noticed. Thanks!

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -1678,25 +1678,40 @@ _Unpickler_SetInputStream(UnpicklerObject *self, PyObject *file)
-                            "file must have 'read' and 'readline' attributes");
+    int no_read_attr = 0;
+    if (_PyObject_LookupAttr(file, &_Py_ID(read), &self->read) < 0) {
+        if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
```
It never matches. `_PyObject_LookupAttr()` converts `AttributeError` into returning value 0.

erlend-aasland replied:
You are correct. I'll adjust the error handling. Thanks!

erlend-aasland replied:
AFAICS, 2d6a7b5e693e4ba0cbe42e0847510454318c2062 should be correct:

- if -1 is returned, clear the exception
- use the `result` output variable instead of temporary `no_read*_attr` variables

serhiy-storchaka reviewed (commented):
I think you missed the purpose `_PyObject_LookupAttr()`. It was designed to get rid of `PyErr_ExceptionMatches(PyExc_AttributeError)` checks and the overhead of raising and catching an exception.

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Address Serhiy's review

- There's no need to match AttributeError, since _PyObject_LookupAttr()
  converts that exception into a return value of 0.
- We can use the result output variable instead of adding temporary
  boolean indicators.

erlend-aasland committed:
Slightly reduce diff

erlend-aasland committed:
Slightly reduce diff

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -1678,25 +1678,30 @@ _Unpickler_SetInputStream(UnpicklerObject *self, PyObject *file)
+        goto error;
+    }
+    if (_PyObject_LookupAttr(file, &_Py_ID(read), &self->read) < 0) {
+        PyErr_Clear();
```
No, we should not replace an arbitrary exception with a TypeError.
```suggestion
        goto error;
```

`PyErr_Clear()` almost always looks like a red herring to me. Normally it should only be called to clear a specified type of exception (after `PyErr_ExceptionMatches()`). If it is called for unspecified type of exception, it is only because we have no way to report an error due to the API limitations (and `PyErr_WriteUnraisable()` is more preferable here), or it does not make much sense to report an error, because we already at abnormal situation handling route.

erlend-aasland replied:
Thanks for the explanation. I would prefer to chain exceptions explicitly in `error`, then.

erlend-aasland replied:
I see now that your code from f320be77ffb73e3b9e7fc98c37b8df3975d84b40 is the best solution. I've updated the PR to be less intrusive. Thanks for your guidance and patience.

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -1678,25 +1678,30 @@ _Unpickler_SetInputStream(UnpicklerObject *self, PyObject *file)
+        PyErr_Clear();
+    }
+    if (_PyObject_LookupAttr(file, &_Py_ID(readline), &self->readline) < 0) {
+        PyErr_Clear();
```
```suggestion
        goto error;
```

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Address Serhiy's review

Don't clear arbitrary exceptions

serhiy-storchaka commented on Modules/_pickle.c:
```diff
@@ -1678,25 +1678,32 @@ _Unpickler_SetInputStream(UnpicklerObject *self, PyObject *file)
-    (void)_PyObject_LookupAttr(file, &_Py_ID(read), &self->read);
-    (void)_PyObject_LookupAttr(file, &_Py_ID(readline), &self->readline);
     if (!self->readline || !self->read) {
         if (!PyErr_Occurred()) {
```
I think that it is better to remove this line, like in your previous version. `_PyObject_LookupAttr()` returns -1 on error, and this case is already handled above.

erlend-aasland replied:
True; we could keep it as an assert.

erlend-aasland replied:
Addressed in b6a8a17e48ef4d69432ed0065b19fb5867933730

erlend-aasland committed:
Pull in main

erlend-aasland committed:
Address Serhiy's review:

There's no need for PyErr_Occurred() anymore
